lib/gcov/ukgcov.c

```diff
@@ -0,0 +1,203 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ * Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the MIT License (the "License", see COPYING.md).
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <string.h>
+#include <uk/essentials.h>
+#include <uk/gcov.h>
+#include <uk/config.h>
+
+#ifdef CONFIG_LIBGCOV_OUTPUT_BINARY_MEMORY
+static unsigned char __used *gcov_output_buffer;
+static size_t __used gcov_output_buffer_pos;
+static size_t gcov_output_buffer_size = INITIAL_BUFFER_SIZE;
+#endif /* CONFIG_LIBGCOV_OUTPUT_BINARY_MEMORY */
+
+/* gcov declarations taken from
+ * https://github.com/gcc-mirror/gcc/blob/master/libgcc/gcov.h
+ */
+struct gcov_info;
+typedef void (*__filename_fn)(const char *, void *);
+typedef void (*__dump_fn)(const void *, unsigned int, void *);
+typedef void *(*__allocate_fn)(unsigned int, void *);
+
+void __gcov_info_to_gcda(const struct gcov_info *__info,
+			 __filename_fn filename_fn, __dump_fn dump_fn,
+			 __allocate_fn allocate_fn, void *__arg);
+void __gcov_filename_to_gcfn(const char *__filename, __dump_fn dump_fn,
+			     void *__arg);
+
+#ifdef CONFIG_LIBGCOV_OUTPUT_BINARY_MEMORY
+static void dump_memory(const void *source, unsigned int n, void *arg __unused)
+{
+	const unsigned char *c = source;
+	unsigned char *buff;
+
+	if (gcov_output_buffer_pos + n >= gcov_output_buffer_size) {
+		buff = (unsigned char *)realloc(gcov_output_buffer,
+						gcov_output_buffer_size <<= 1);
+
+		if (unlikely(!buff)) {
+			free(gcov_output_buffer);
+			return;
+		}
+		gcov_output_buffer = buff;
+	}
+	memcpy(gcov_output_buffer + gcov_output_buffer_pos, c, n);
+	gcov_output_buffer_pos += n;
+}
+#endif /* CONFIG_LIBGCOV_OUTPUT_BINARY_MEMORY */
+
+#ifdef CONFIG_LIBGCOV_OUTPUT_BINARY_FILE
+static void dump_file(const void *source, unsigned int n, void *arg)
+{
+	FILE *file = arg;
+	const unsigned char *c = source;
+
+	fwrite(c, n, 1, file);
+}
+#endif /* CONFIG_LIBGCOV_OUTPUT_BINARY_FILE */
+
+#ifdef CONFIG_LIBGCOV_OUTPUT_SERIAL_HEXDUMP
+static inline char e(unsigned int c)
+{
+	return (c % 16) < 10 ? (c % 16) + '0' : (c % 16) - 10 + 'a';
+}
+
+static inline unsigned char *encode(unsigned char c, unsigned char buf[2])
+{
+	buf[0] = e(c);
+	buf[1] = e(c >> 4);
+	return buf;
+}
+
+static void dump_serial(const void *source, unsigned int n, void *arg __unused)
+{
+	const unsigned char *c = source;
+	unsigned char buf[2];
+
+	for (unsigned int i = 0; i < n; ++i) {
+		char *encoded = encode(c[i], buf);
```

Comment:
According to the coding conventions we generally avoid declaring block scope vars, so it's perhaps better to move the declaration of `char *encoded` to the top.